<!doctype html>

<html lang="en">
<head>
  <meta charset="utf-8">
	<meta name="robots">
  <title>The FEST analysis user's manual</title>

</head>

<body style="background-color:#e6ffcc;">
<p><span style="font-size: 16pt; font-weight: 700"><a name="top"></a>How to work with the FEST web application</span></p>
<p>This article contains the following sections:</p>

<ul>
	<li><a href="#Overview">Overview</a></li>
	<li><a href="#inputFiles">Input files</a></li>
	<li><a href="#inputParams">Input parameters</a></li>
	<li><a href="#outputParams">Output</a></li>
	<li><a href="#ref">Reference</a></li>
</ul>

		<h2><a name="Overview">Overview</a></h2>
		<p>The <a href="http://www.stat-apps.onc.jhmi.edu/FEST/">FEST</a> (Functional Expansion of Specific T-cells) platform is a Shiny application 
		for analysis of TCR sequencing of short-term, 
		peptide-stimulated cultures to identify antigen-specific clonotypic amplifications. The app loads TCR sequencing data in the tab-delimited format,
 performs the analysis, and visualizes and saves results. For analysis, we use only productive clones and
 summarize template counts for nucleotide sequences that translated into the same amino acid sequence.
 For each analyzed clone, we apply Fisherâ€™s exact test to compare the number of templates in a culture of interest
 (with peptide) and a reference culture (T cells cultured with cytokines but without peptide). The p-value adjusted by Benjamini-Hochberg
 procedure (FDR) is used to determine antigen-specific clonotypes (FEST positive clones) that
 met the following criteria: 
 <ol>
	 <li>significantly expanded in the culture of interest compared to the reference
	 culture at an FDR less than the specified threshold
	 (a value from the <b>Specify FDR threshold</b> field), <br>
	 <b>NOTE:</b> This step is omitted if the <b>Compare to reference</b> option is disabled</li>
	 <li>significantly expanded in the culture of interest compared to every other culture
	 performed in tandem (a value from the <b>Specify FDR threshold</b> field),</li> 
	 <li>have an odds ratio above a value specified in the <b>Specify odds ratio threshold</b> field, </li>
	 <li>a minimum template threshold in uncultured T cells (baseline) calculated by:<br>
	 limit = 1-(1-P)<sup> (1/n)</sup><br>
	 Where P = the probability of observing the clone in a given well (clone confidence specified in the <b>Specify clone confidence</b> field) and 
	 n = the estimated number of CD8+ T cells per well prior to culture (specified in the <b>Estimated number of CD8+ T cells per well</b> field).  <br>
	 <b>NOTE:</b> This threshold is not applied if the <b>Ignore baseline threshold</b> option is enabled</li>
 </ol>
 Clones can then be subject to a threshold for a minimal number of templates (the <b>Specify the minimal number of templates</b> field). FEST positive clones are
 saved in the output table (the <b>Download Results</b> button) and plotted as an output heat map (the <b>Create Heat Map</b> button).</p>

<a href="#top"><font size="2">Top of Page</font></a></P>

<h2><a name="inputFiles">Input files</a></h2>
<p> First, you need to specify a format of your input file(s). The app allows you to upload TCR sequencing data in the tab-delimited format or .rda file that were previously saved from the app. 
			Currently, the app supports two formats of TCR sequencing data in the tab-delimited format:
		<ul>
			<li>files exported from <a href="https://www.adaptivebiotech.com/immunoseq">Adaptive Biotechnologies ImmunoSEQ</a> platform <br>
			</li>
			<li>files exported from <a href="https://vdjtools-doc.readthedocs.io/en/master/intro.html">VDJtools</a> </li>
			<b>NOTE:</b> Multiple file selection does not work on older browsers, including Internet Explorer 9 and earlier.
		</ul>
		</p>
		<a href="#top"><font size="2">Top of Page</font></a>

<h2><a name="inputParams">Input parameters</a></h2>
	<p>Before running the analysis, you need to specify the following input parameters:
	
	<ul>
		<li> <b>Compare to reference</b>. If you want to omit the comparison to a reference sample step, disable this option.</li>
		<li> <b>Select a reference sample</b>. It is usually T cells cultured with cytokines but without peptide.</li>
		<li> <b>Select a baseline sample</b>. It is uncultured T cells. This sample will be used to calculate a minimum template threshold for clones in uncultured T cells. 
		Only clones that passed this threshold will be analyzed. If you do not want to apply this threshold, select the <b> Ignore baseline threshold</b> checkbox. If you do not specify a baseline sample, the reference sample will be used to calculate the baseline threshold</li>
		<li> <b>Select sample(s) to exclude</b>. The samples will be excluded from analysis, but information about them will be included in the output file</li>
		<li> <b>Specify the minimal number of templates</b>. A threshold for the number of templates in cultures of interest</li>
		<li> <b>Ignore baseline threshold</b>. If this option is selected, only <b>the minimal number of templates</b> threshold will be used. If you select this option, we recommend you to increase the minimal number of templates to reduce the number of clones to analyze. Smaller number of templates leads to bigger number of clones to analize and slows down the analysis</li>
		<li> <b>Specify clone confidence</b>. The confidence will be used to calculate a minimum template threshold for clones in uncultured T cells, 
			if <b>Ignore baseline threshold</b> is selected. The lower the confidence, the more clones will be analyzed</li>
		<li> <b>Estimated number of CD8+ T cells per well</b>. This number will be used to calculate a minimum template threshold for clones in uncultured T cells, 
			if <b>Ignore baseline threshold</b> is selected.</li>
		<li> <b>Use nucleotide level data</b>. Select this option if you want to perform analysis on the nucleotide level data instead of amino acid level data. 
			By default, the analysis is performed on amino acid level.</li>
	</ul>
	After setting all parameters, press the <b>Run Analysis</b> button.
	
	</p>
	<a href="#top"><font size="2">Top of Page</font></a>

<h2><a name="outputParams">Output</a></h2>
		<p>	After the analysis is done, you can specify FDR and odds ratio thresholds that will be used to determine FEST positive clones. 
		The positive odds ratio corresponds to the expansion of a clone in cultures of interest. 
		The results are saved by clicking the <b>Download Results</b> button in an Excel file with the following spreadsheets:
		<ul>
			<li> The <b>condition_summary</b> sheet contains the number of positive clones and the sum of percents of the clones per condition </li>
			<li> The <b>positive_clones_summary</b> sheet contains a list of positive clones, condition in which a clone is expanded, percent of a clone 
				in the baseline and the reference samples, and the fold change of a clone comparing to the baseline and the reference samples</li>
			<li> The <b>positive_clones_all_data</b> sheet contains extended information about the positive clones including 
			the number of templates (the '*_abundance' columns) and the percent (the '*_percent' columns)
			in all samples including those that were excluded from analysis 

			<li> The <b>ref_comparison_only</b> sheet contains a list of all analyzed clones, FDRs (the 'FDR:*' columns) and odds ratios (the 'OR:*' columns) for comparisons between cultures of interest 
				and the reference sample, as well as the number of templates (the '*_abundance' columns) and the percent (the '*_percent' columns)
				in all samples including those that were excluded from analysis<br>
				<b>NOTE:</b> This sheet is omitted if the <b>Compare to reference</b> option is disabled</li>
			<li> The <b>parameters</b> sheet contains all input parameters, the number of samples analyzed, and the total number of templates 
			corresponding to productive clones per sample</li>
		</ul>
		You also can save the positive clones as a heat map by clicking the <b>Create Heat Map</b> button.
		</p>
	<a href="#top"><font size="2">Top of Page</font></a></P>

<h2><a name="ref">Reference</a></h2>
	<p>The manuscript is published in <a href="http://cancerimmunolres.aacrjournals.org/content/early/2018/06/12/2326-6066.CIR-18-0129" target="_blank" >Cancer Immunology Research.</a></p>
	To cite the paper, please use the following reference:<br>
	The Mutation-Associated Neoantigen Functional Expansion of Specific T cells (MANAFEST) assay: a sensitive platform for monitoring antitumor immunity<br>
Ludmila Danilova, Valsamo Anagnostou, Justina X Caushi, John-William Sidhom, Haidan Guo, Hok Yee Chan, Prerna Suri, Ada J. Tam, Jiajia Zhang, Margueritta El Asmar, Kristen A Marrone, Jarushka Naidoo, Julie R. Brahmer, Patrick M Forde, Alexander S. Baras, Leslie Cope, Victor E. Velculescu, Drew Pardoll, Franck Housseau and Kellie N. Smith
<br>Cancer Immunol Res June 12 2018 DOI: 10.1158/2326-6066.CIR-18-0129

	<p>This shiny app is publicly available at <a href="http://www.stat-apps.onc.jhmi.edu/FEST/">http://www.stat-apps.onc.jhmi.edu/FEST/</a>
and the source code has been deposited at <a href="https://sourceforge.net/projects/manafest/">https://sourceforge.net/projects/manafest/</a></p>

	<a href="#top"><font size="2">Top of Page</font></a></P>
</body>
</html>
